genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/grpc/GRpcAgentFileManifestServiceImpl.java

```diff
@@ -0,0 +1,157 @@
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+
+package com.netflix.genie.agent.execution.services.impl.grpc;
+
+import com.google.protobuf.Empty;
+import com.netflix.genie.agent.execution.services.AgentFileManifestService;
+import com.netflix.genie.common.internal.dto.JobDirectoryManifest;
+import com.netflix.genie.common.internal.dto.v4.converters.JobDirectoryManifestProtoConverter;
+import com.netflix.genie.common.internal.exceptions.GenieConversionException;
+import com.netflix.genie.common.internal.util.ExponentialBackOffTrigger;
+import com.netflix.genie.proto.JobFileManifestMessage;
+import com.netflix.genie.proto.JobFileServiceGrpc;
+import io.grpc.stub.StreamObserver;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.scheduling.TaskScheduler;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.concurrent.ScheduledFuture;
+
+/**
+ * Implementation of {@link AgentFileManifestService} that publishes the manifest over gRPC.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Slf4j
+public class GRpcAgentFileManifestServiceImpl implements AgentFileManifestService {
+    private final JobFileServiceGrpc.JobFileServiceStub jobFileServiceStub;
+    private final TaskScheduler taskScheduler;
+    private final ExponentialBackOffTrigger trigger;
+    private final JobDirectoryManifestProtoConverter manifestProtoConverter;
+    private StreamObserver<JobFileManifestMessage> manifestStreamObserver;
+    private String jobId;
+    private File jobDirectory;
+    private boolean started;
+    private ScheduledFuture<?> scheduledTask;
+    private final StreamObserver<Empty> responseObserver;
+
+    GRpcAgentFileManifestServiceImpl(
+        final JobFileServiceGrpc.JobFileServiceStub jobFileServiceStub,
+        final TaskScheduler taskScheduler,
+        final JobDirectoryManifestProtoConverter manifestProtoConverter
+    ) {
+        this.jobFileServiceStub = jobFileServiceStub;
+        this.taskScheduler = taskScheduler;
+        this.manifestProtoConverter = manifestProtoConverter;
+        this.trigger = new ExponentialBackOffTrigger(
+            ExponentialBackOffTrigger.DelayType.FROM_PREVIOUS_EXECUTION_BEGIN,
+            1000, //TODO make configurable
+            10000, //TODO make configurable
+            1.1f //TODO make configurable
+        );
+        this.responseObserver = new ManifestResponseObserver(this);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public synchronized void start(final String claimedJobId, final File jobDirectoryRoot) {
+        //Service can be started only once
+        if (started) {
+            throw new IllegalStateException("Service can be started only once");
+        }
+
+        this.jobId = claimedJobId;
+        this.jobDirectory = jobDirectoryRoot;
+        this.started = true;
+        this.scheduledTask = this.taskScheduler.schedule(this::pushManifest, trigger);
```

Comment:
You could also just use the `@Scheduled` annotation on the method. The only problem with that is you lose some control

Reply:
Right, i want to be able to cancel, and it's easier to mock if i schedule manually.